### PR TITLE
Container is now considered dirty if not attached.

### LINKED ISSUE #8358: container.isDirty flag should be true for detached containers
I'm working on showing users a warning dialog when they are navigating away from a page that is rendering a Fluid container that has pending ops. To achieve this, I use the isDirty flag and the 'saved' event. This covers the case where we have an attached container that has not finished pushing ops to the server.

One other scenario we'd like to handle is a container that is not yet attached. Refreshing in this case would also cause data loss, so we should also warn the user in this case. Ideally, we'd be able to use the same isDirty and 'saved' APIs to achieve this as well. Right now, isDirty is false for detached containers based on local testing. My proposal is that the isDirty flag should be true for containers that have not had attach called or have an attach call in flight. Once the container is attached and all ops are persisted, it should set isDirty to false and emit a saved event.

### PR DESCRIPTION
**Pull request by ssimic2**
Closes https://github.com/microsoft/FluidFramework/issues/8358.

Container is now considered "dirty" if it is NOT attached or there are pending ops. Generally, "dirty" may not be the best way to describe this state, rather saved (vs not saved) would have probably been more appropriate.

### CODE REVIEW

**Comment by vladsud** [comment]
> The logic does not take into account any DataStoreRuntime ops (like creating channels etc.) So, as long as the Container is not attached, the flag will always be "dirty".

There are no ops if container is detached. There could be ops if container is attaching or attached.

File: docs/content/docs/build/containers.md [hunk 1]
```diff
@@ -149,8 +149,9 @@ You should always check the `isDirty` flag before closing the container or navig
 
 A container is considered dirty in the following cases:
 
-1. The container has local changes that have not yet been acknowledged by the service. These unacknowledged changes will be lost if the container is closed.
-1. There is no network connection while making changes to the container. These changes cannot be acknowledged by the service until the network connection is restored.
+1. The container has been created in the detached state, and it has not been attached yet.
+2. The container was attached, but it has local changes that have not yet been acknowledged by the service. These unacknowledged changes will be lost if the container is closed.
```
Comment (vladsud) [inline docs/content/docs/build/containers.md:153]: will be lost -> maybe lost
It's not certain, maybe ops were sent, but we simply do not know if they were received.

File: docs/content/docs/build/containers.md [hunk 2]
```diff
@@ -149,8 +149,9 @@ You should always check the `isDirty` flag before closing the container or navig
 
 A container is considered dirty in the following cases:
 
-1. The container has local changes that have not yet been acknowledged by the service. These unacknowledged changes will be lost if the container is closed.
-1. There is no network connection while making changes to the container. These changes cannot be acknowledged by the service until the network connection is restored.
+1. The container has been created in the detached state, and it has not been attached yet.
```
Comment (vladsud) [inline docs/content/docs/build/containers.md:152]: Given that (I think) we expose tri-state (attached, attaching, detached) I think this description has to talk about attaching state. In particular - that container is dirty in this state, and that if container is closed, host may never know if file was created or not.

File: docs/content/docs/build/containers.md [hunk 3]
```diff
@@ -149,8 +149,9 @@ You should always check the `isDirty` flag before closing the container or navig
...
+2. The container was attached, but it has local changes that have not yet been acknowledged by the service. These unacknowledged changes will be lost if the container is closed.
+3. There is no network connection while making changes to the container. These changes cannot be acknowledged by the service until the network connection is restored.
```
Comment (vladsud) [inline docs/content/docs/build/containers.md:154]: Isn't 3 a subset of 2? I'm not sure if we want to say that (maybe we do).
I think one think you can add - that state will flicker quickly between dirty and saved as changes are made and that host may choose to use some delay before showing that in UX - both as continuous status update, as well as raising "are you sure you want to close this doc? You have unsaved changes" kind of dialog on close.

File: packages/loader/container-loader/src/container.ts [hunk 4]
```diff
@@ -1837,7 +1837,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
     ) {
         assert(this._context?.disposed !== false, 0x0dd /* "Existing context not disposed" */);
         // If this assert fires, our state tracking is likely not synchronized between COntainer & runtime.
-        if (this._dirtyContainer) {
+        if (!this._dirtyContainer) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1840]: I'd think the purpose of this check was - log that we are losing user data if we are forcing to reload runtime.
I.e., it's less about first runtime usage (likely reason you are hitting this error and changing condition), but about any further reloads and losing user data because of it.
That said, when I looked at usage of instantiateContext(), sounds like it's used only when we boot container. If so (please validate), then I'd rather remove this check and error telemetry.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -875,7 +875,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     private _disposed = false;
     public get disposed() { return this._disposed; }
 
-    private dirtyContainer = false;
+    private dirtyContainer = true;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:878]: I think the intention is - constructor will / has to set right value. If so, I'd remove initialization part here and leave only declaration, to help compiler catch bugs if constructor would not set it up properly

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 6]
```diff
@@ -1780,6 +1782,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             this.emit("attached");
         }
         this.dataStores.setAttachState(attachState);
+        this.resetDirtyState(attachState);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1785]: I do not think this call here is correct when we transition to "attached" state. Container maybe dirty, maybe not - it depends on whether it has pending ops or not.
Prior to that (detached and attaching state) it's dirty - that's fine.

File: packages/test/test-utils/src/loaderContainerTracker.ts [hunk 7]
```diff
@@ -167,8 +167,10 @@ export class LoaderContainerTracker implements IOpProcessingController {
             if (containersToApply.length === 0) { break; }
 
             // Ignore readonly dirty containers, because it can't sent up and nothing can be done about it being dirty
-            const dirtyContainers =
-                containersToApply.filter((c) => c.deltaManager.readOnlyInfo.readonly !== true && c.isDirty);
+            const dirtyContainers = containersToApply.filter((c) => {
+                const { deltaManager, attachState,isDirty } = c;
+                return deltaManager.readOnlyInfo.readonly !== true && attachState === AttachState.Attached && isDirty;
```
Comment (vladsud) [inline packages/test/test-utils/src/loaderContainerTracker.ts:172]: I'm not sure why this change is necessary and why / if it is correct.
I'd think that usage of this API is limited to only attached containers, i.e., tests should ensure attachment process is finished before calling this API. If not, there is only one container (with unfinished file creation) to sync and calling this API and it returning right away does not give caller anything useful.

File: docs/content/docs/build/containers.md (hunk 1 above)
Comment (ssimic2) [reply docs/content/docs/build/containers.md:153]: Good point. This is some older language there. Not getting an ack back does not mean the service did not receive "op".

File: docs/content/docs/build/containers.md (hunk 2 above)
Comment (ssimic2) [reply docs/content/docs/build/containers.md:152]: I can call out "attaching" state here explicitly as part of 1. The message here should probably still be concise and boil down to, if container is not  in "attached" state (whether detached or attaching), it is "dirty".

File: docs/content/docs/build/containers.md (hunk 3 above)
Comment (ssimic2) [reply docs/content/docs/build/containers.md:154]: Agree, its a subset. This was an older language that differentiated the two - will make the change.

Interesting point about flickering. Are there UX use cases where we expose "saved"/"dirty" state continuously to user as they modify shared data? Id imagine we would still have quite a bit of flickering even in normal ops flow. Or these are mostly one-offs, for example on document close like you mentioned ?

File: packages/loader/container-loader/src/container.ts (hunk 4 above)
Comment (ssimic2) [reply packages/loader/container-loader/src/container.ts:1840]: @vladsud good to know this. I'll double check it. I think I had one test case fail on this check.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 6 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1785]: I just re-run couple of test cases again, and you are right about not needing to make this call on attach. Interestingly, there are no ops when container is attached. Meaning, the container summary is sent out without any `attach` ops. What triggers the reset of this flag is connection to delta manager (that coincidentally happens on  `attach`). Runtime tries to replay states (again, there are no pending ops), and subsequently clears the dirty flag. @vladsud Is this what you expected? In one way, it does make sense to flush on "connect", but I do find attach/connect/bind flow somewhat complex to reason about, and less cohesive. I'm still learning about it though :)

Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Transitioning to attached state means file was created. At that moment we have file identity and can connect to relay service to send any ops. We might have some ops to send (that were generated while container was in attaching state) or not - that's not give. Just want to make sure it's clear that switch to attached state does not mean container is not dirty - it depends ("there are no pending ops" is incorrect in this context).
But I have not followed all the logic - it's possible it's fine to set it to be not dirty with assumption that logic that fires right after that due to "connected" event/transition will properly transition state to dirty if needed. That said, even if that's the case, it seems fishy that we would transition to "saved" state even for a blink of an eye, with pending changes. If possible, I'd prefer for "saved" state to be exposed to consumers only when there are truly no pending changes.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 8]
```diff
@@ -1078,6 +1078,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
 
         this.summaryCollection = new SummaryCollection(this.deltaManager, this.logger);
 
+        this.dirtyContainer = this.context.attachState !== AttachState.Attached;
+        this.context.updateDirtyContainerState(this.dirtyContainer);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1082]: @wes-carlson , can you please glance? I believe we should set container dirty if this.getPendingLocalState() !== undefined, but I do not see any usage in the code of getPendingLocalState(). Still might be worth bullet-proofing code for future.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 6 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Thanks for this clarification. I'll resolve this convo (as this call was removed) and look at the below comment around the flag init in constructor, keeping this use case in mind. Agree, we should be consistent: attached + pending ops = dirty (with no exceptions).

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1082]: Only pending ops that can come in through runtime constructor are via `context.pendingLocalState` that gets subsequently passed to `PendingStateManager`. However, PendingStateManager does not consider that set when asked about `hasPendingMessages`. I'll dig in further in the morning.

File: packages/loader/container-loader/src/container.ts (hunk 4 above)
Comment (ssimic2) [reply packages/loader/container-loader/src/container.ts]: Removed. We do `instantiateContext` right after we boot the container.

File: packages/test/test-utils/src/loaderContainerTracker.ts (hunk 7 above)
Comment (ssimic2) [reply packages/test/test-utils/src/loaderContainerTracker.ts:172]: I'll retest this after dirty flow correction. If isDirty is decoupled from Attach logic, this will probably go away.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1082]: @wes-carlson I'd appreciate your take here. It seems that we can create runtime either in detached state (in which case we don't have any pending ops) or attached state where `context.pendingLocalState` will list out stashed ops that need to be moved to pending state. If we want to be very accurate on `dirty` state we probably need to consider both `context.attachState` and  `context.pendingLocalState`.  

However, within runtime, both logic flows (detached and attached) funnel through replaying pending states when `join` triggers the connection state change, at which point dirty flag is correctly cleared based on the fact that all pending ops just have been sent through. So alternatively, we could default runtime to dirty state (in the constructor), and let the connection event flow clear the flag.

Maybe we could chat briefly on use cases I missed here (ex. implications of read-only to r/w transitions).

**Comment by tylerbutler** [comment]
@ssimic2 The docs wording looks OK, but I am going to wait to do another pass once you've figured out some of the nuances with Wes, in case some of those details need to be included in the docs.

**Comment by anthony-murphy** [comment]
this doesn't exist today, but i could see a detached container being dirty vs saved, as we allow serialization of the detached container, so a user may want to know if there are changes since the last time it serialized. given this, we may not want to couple attach state and dirty/saved

i'd like to see an issue here with some level of design discussion

**Comment by vladsud** [comment]
> this doesn't exist today, but i could see a detached container being dirty vs saved, as we allow serialization of the detached container, so a user may want to know if there are changes since the last time it serialized. given this, we may not want to couple attach state and dirty/saved

It's less about detached state (as we build serialization and rehydration for both attached & detached states to preserve user data). You are right, that host may want to know if anything changed since last serialization.
As of now, hosts do not need this capability. I.e., no host wants to periodically save document state and know if it needs to update their save. They all want to save document state only on user moving away. So, I propose we do not add yet that complication, and clearly document what dirty means in current system.

**Comment by ssimic2** [comment]
> this doesn't exist today, but i could see a detached container being dirty vs saved, as we allow serialization of the detached container, so a user may want to know if there are changes since the last time it serialized. given this, we may not want to couple attach state and dirty/saved
> 
> i'd like to see an issue here with some level of design discussion

Hey @anthony-murphy I'd love to hear more about these use cases, and if anyone has needed /requested them. It's OK if its not pressing need, I'm curious how some of these use cases do evolve over time. Is it us offering "you can also do this" or its something that's coming through partners.

Some general thoughts on APIs that are surfaced to end users (as this one in) ... We should be very cognizant of the mental model that we form. In other words, we should be able to reason about API fairly quickly, without (whenever possible) needing to link into further concepts/APIs, or/and explain contextual terminology. In my head, when I onboarded, `isDirty` being `false` meant the doc is live (aka `attached`) and saved (no pending ops). All concrete/known concepts, and no further explanations were needed to reason about what `isDirty` meant. Name could have been better (maybe `saved`), but I understood intention. Of course over time you learn about other concepts (like readonly vs rw containers), but as long as these concepts map to state/ops (the terminology which we used to explain `isDirty`) then we don't need them in the context of our flag explanation. For all API changes that may introduce additional reasoning/concepts behind meaning of `isDirty` we should have a high bar and let those come in only when there is a highly significant value. And again this is general thought, the `serialize` use case may or may not land into this bucket.

I'll check with Wes on load nuances to see if we can handle transient state more gracefully here.

**Comment by anthony-murphy** [comment]
> > this doesn't exist today, but i could see a detached container being dirty vs saved, as we allow serialization of the detached container, so a user may want to know if there are changes since the last time it serialized. given this, we may not want to couple attach state and dirty/saved
> > i'd like to see an issue here with some level of design discussion
> 
> Hey @anthony-murphy I'd love to hear more about these use cases, and if anyone has needed /requested them. It's OK if its not pressing need, I'm curious how some of these use cases do evolve over time. Is it us offering "you can also do this" or its something that's coming through partners.
> 
> Some general thoughts on APIs that are surfaced to end users (as this one in) ... We should be very cognizant of the mental model that we form. In other words, we should be able to reason about API fairly quickly, without (whenever possible) needing to link into further concepts/APIs, or/and explain contextual terminology. In my head, when I onboarded, `isDirty` being `false` meant the doc is live (aka `attached`) and saved (no pending ops). All concrete/known concepts, and no further explanations were needed to reason about what `isDirty` meant. Name could have been better (maybe `saved`), but I understood intention. Of course over time you learn about other concepts (like readonly vs rw containers), but as long as these concepts map to state/ops (the terminology which we used to explain `isDirty`) then we don't need them in the context of our flag explanation. For all API changes that may introduce additional reasoning/concepts behind meaning of `isDirty` we should have a high bar and let those come in only when there is a highly significant value. And again this is general thought, the `serialize` use case may or may not land into this bucket.
> 
> I'll check with Wes on load nuances to see if we can handle transient state more gracefully here.

This all sounds right to  me. I wonder if we should just introduce the concept of saved which is what you allude to here, and leave dirty as is. I've seen a few people get tripped up by the word dirty, so moving to a better concept would be worth it. We could even consider deprecating dirty if we don't think its valuable as an independent concept

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1082]: I think it makes sense to look at context.pendingLocalState. I don't think we'd want to default to dirty since if there are no stashed ops it will be incorrect until `join` happens.

File: packages/test/test-utils/src/loaderContainerTracker.ts (hunk 7 above)
Comment (ssimic2) [reply packages/test/test-utils/src/loaderContainerTracker.ts]: This change has been reverted.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1085]: Now including `context.pendingLocalState` in addition to attached state. In conclusion, the runtime construction logic:
* Detached context -> Dirty 
* Attached context + pending ops -> Dirty 
* Attached context + no pending ops -> Not Dirty 

`local-server-tests\src\test\documentDirty.spec.ts` && `test-end-to-end-tests\src\test\containerDirtyFlag.spec.ts` should validate all the use cases, inclusive of the ones we discussed here.

**Comment by ssimic2** [comment]
> > > this doesn't exist today, but i could see a detached container being dirty vs saved, as we allow serialization of the detached container, so a user may want to know if there are changes since the last time it serialized. given this, we may not want to couple attach state and dirty/saved
> > > i'd like to see an issue here with some level of design discussion
> > 
> > 
> > Hey @anthony-murphy I'd love to hear more about these use cases, and if anyone has needed /requested them. It's OK if its not pressing need, I'm curious how some of these use cases do evolve over time. Is it us offering "you can also do this" or its something that's coming through partners.
> > Some general thoughts on APIs that are surfaced to end users (as this one in) ... We should be very cognizant of the mental model that we form. In other words, we should be able to reason about API fairly quickly, without (whenever possible) needing to link into further concepts/APIs, or/and explain contextual terminology. In my head, when I onboarded, `isDirty` being `false` meant the doc is live (aka `attached`) and saved (no pending ops). All concrete/known concepts, and no further explanations were needed to reason about what `isDirty` meant. Name could have been better (maybe `saved`), but I understood intention. Of course over time you learn about other concepts (like readonly vs rw containers), but as long as these concepts map to state/ops (the terminology which we used to explain `isDirty`) then we don't need them in the context of our flag explanation. For all API changes that may introduce additional reasoning/concepts behind meaning of `isDirty` we should have a high bar and let those come in only when there is a highly significant value. And again this is general thought, the `serialize` use case may or may not land into this bucket.
> > I'll check with Wes on load nuances to see if we can handle transient state more gracefully here.
> 
> This all sounds right to me. I wonder if we should just introduce the concept of unsaved which is what you allude to here, and leave dirty as is. I've seen a few people get tripped up by the word dirty, so moving to a better concept would be worth it. We could even consider deprecating dirty if we don't think its valuable as an independent concept

Yes, @ChumpChief called it something very similar. So maybe `isUnsaved` as a flag. @vladsud @tylerbutler your thoughts? I would not have two different behaviors though. It would add unnecessary confusion. I'd say we move this current change in, and if we decide to rename the flag, let's handle that through another ticket that will just deal with deprecation/alternative API, essentially renaming identical behavior.

**Comment by vladsud** [comment]
I think whatever name we chose; it will have roughly same outcome - someone will assume different behavior.
That said, "synchronized" seems like a better term if we really want to point to a fact that it's about state in the cloud (though synchronized assumes bi-directional communication, whereas here we only assess one way.
Apps use "saving" & "saved" in UX for same thing ("saving" might be deceiving in offline as it assumes active process, but that's what we use)

**Comment by ssimic2** [comment]
> I think whatever name we chose; it will have roughly same outcome - someone will assume different behavior. That said, "synchronized" seems like a better term if we really want to point to a fact that it's about state in the cloud (though synchronized assumes bi-directional communication, whereas here we only assess one way. Apps use "saving" & "saved" in UX for same thing ("saving" might be deceiving in offline as it assumes active process, but that's what we use)

All good points @vladsud. I would say, lets wrap up the requested changes this PR is targeting and take this as a separate point then. All earlier comments are resolved, so please give it another look when you get a chance.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -1079,6 +1079,11 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
 
         this.summaryCollection = new SummaryCollection(this.deltaManager, this.logger);
 
+        const { attachState, pendingLocalState } = this.context;
+        this.dirtyContainer = attachState !== AttachState.Attached
+            || (pendingLocalState as IPendingLocalState)?.pendingStates.length > 0;
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1084]: The calls back and forth between the context/runtime on pending local state are confusing to me, e.g. I don't understand why we wouldn't directly call `this.getPendingLocalState()`.  Maybe this area (pending state) needs some more attention in general in the future?

Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1084]: Good points @ChumpChief. Interestingly, `PendingStateManager` does not consider this initial state (`pendingLocalState` coming from context) as pending ops. Only at a later time, when stashed ops are applied at their reference sequence is when initial state transitions to local state within `PendingStateManager`, and is actually considered pending. So, requesting pending ops count on the initialized `PendingStateManager` in the runtime constructor would just give back 0, even though `stashed` ops is present. I was thinking well let's expose this added API on pending manager (like getStashedOps or getShashedOpsCount), but then we introduce another term and another API with somewhat limited purpose.

Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1084]: @wes-carlson .... if you have any thoughts on the above.

Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1084]: @wes-carlson if you think Pending Manager is work in progress (Im not sure what further complexities you are thinking about), do you think making related pending manager changes is appropriate to do it as part of this PR (if so, happy to do it here). Or keep things as they are in this PR and rely on context as done above ?

Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1084]: I'll leave the pending manager change out of this PR then, as it needs a bit more work there unrelated to dirty flag. From the perspective of runtime constructor, looking at `pendingLocalState` is enough to (correctly) determine there are pending ops present.

File: docs/content/docs/build/containers.md [hunk 10]
```diff
@@ -143,14 +143,14 @@ container.on("disposed", () => {
...
-
-You should always check the `isDirty` flag before closing the container or navigating away from the page.
+A container is considered **dirty** if it has local changes that have not yet been acknowledged by the service. You should always check the `isDirty` flag before closing the container or navigating away from the page. Closing the container while `isDirty === true` may result in the loss of operations that have not yet been acknowledged by the service.
 
 A container is considered dirty in the following cases:
 
-1. The container has local changes that have not yet been acknowledged by the service. These unacknowledged changes will be lost if the container is closed.
-1. There is no network connection while making changes to the container. These changes cannot be acknowledged by the service until the network connection is restored.
+1. The container has been created in the detached state, and either it has not been attached yet or it is in the process of being attached (container is in `attaching` state). Changes are not sent to the service until the container is *attached*. If container is closed prior to being attached, host may never know if the file was created or not.
```
Comment (vladsud) [inline docs/content/docs/build/containers.md:150]: > Changes are not sent to the service until the container is *attached*

I'd remove this. I think people will not follow what you want to say here. If you really want to stress it, I'd say that "container may stay in dirty state when transitioning from attaching to attached state due to changes made in attaching state". But I do not know if we want to put all these complexities here.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1084]: I think it's worth opening an issue to follow up here to not have such cast in the future

**Review by vladsud — APPROVED** [review]

File: docs/content/docs/build/containers.md (hunk 10 above)
Comment (ssimic2) [reply docs/content/docs/build/containers.md]: Yes, removed - that paragraph would make sense without that sentence as well. I like the idea of simplifying it.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1084]: I combined coupe of items we discussed around pending states and pending state manager here: https://github.com/microsoft/FluidFramework/issues/8893

**Comment by ssimic2** [comment]
> @ssimic2 The docs wording looks OK, but I am going to wait to do another pass once you've figured out some of the nuances with Wes, in case some of those details need to be included in the docs.

@tylerbutler this PR is a bit old, but it should be ready to merge after the build goes through. Please give it another pass, and let me know if you have any further thoughts. I'd love to hear your take on the naming conversation we had with @anthony-murphy and @vladsud above. I'll open another ticket that will focus on renaming (of existing behavior) only.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1086]: https://github.com/microsoft/FluidFramework/issues/8893

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1086]: Can we create an issue to track the work in pending state manager?

**Comment by ssimic2** [comment]
> > @ssimic2 The docs wording looks OK, but I am going to wait to do another pass once you've figured out some of the nuances with Wes, in case some of those details need to be included in the docs.
> 
> @tylerbutler this PR is a bit old, but it should be ready to merge after the build goes through. Please give it another pass, and let me know if you have any further thoughts. I'd love to hear your take on the naming conversation we had with @anthony-murphy and @vladsud above. I'll open another ticket that will focus on renaming (of existing behavior) only.

https://github.com/microsoft/FluidFramework/issues/8960

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1087]: Can we add this as a comment where we initialize the dirty state. It will be super helpful and clear :)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 11]
```diff
@@ -1783,6 +1788,10 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
                 0x12e /* "Container Context should already be in attached state" */);
             this.emit("attached");
         }
+
+        if (attachState === AttachState.Attached && !this.pendingStateManager.hasPendingMessages()) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1792]: We don't care about the local state passed to the pending state manager here?

**Comment by agarwal-navin** [comment]
A note should be added to the BREAKING.md file as this is change in behavior.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 11 above)
Comment (ssimic2) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1792]: I have that note here: https://github.com/microsoft/FluidFramework/issues/8893. I'll sort out the pending (manager) ticket as a follow up in Feb. However, realistically with use cases we have, this is a minor issue at this point.

**Review by agarwal-navin — APPROVED** [review]
lgtm. Only thing I am not sure about is the interaction with the pending local state during pending state manager initialization but it looks like there is conversation with Wes about it.

**Review by tylerbutler — APPROVED** [review]
